### PR TITLE
Add ApplicationSummaryInfo wrapper to allow mock tests

### LINKED ISSUE #6707: Add unit tests to the AutoTuner that rely on app eventlogs
Currently, it is hard to add unit tests for Tools.
The main reason is that some features need input from processing eventlogs.

After discussion with @tgravescs  we agreed to have the following refactor:

- Implement a wrapper that consumes `ApplicationSummaryInfo`
- Change the constructor of the AutoTuner to accept an interface/treat to get different metrics from the app
- In the unit tests, build mocks that directly provide those functions.

This should be the future direction for testing the Tools functionality.

### LINKED ISSUE #6895: [BUG] Profile tool fails in getMaxTaskInputSizeBytes
**Describe the bug**
I found that when running the profiler tool on an event log, Analysis.scala `getMaxTaskInputSizeBytes()` crashes applying max on an empty array.
In this case, the profiling for the given app generates no output. 

```
22/10/24 12:28:47 WARN Profiler: Exception occurred processing file: profile_testing
java.lang.UnsupportedOperationException: empty.max
	at scala.collection.TraversableOnce.max(TraversableOnce.scala:275)
	at scala.collection.TraversableOnce.max$(TraversableOnce.scala:273)
	at scala.collection.AbstractTraversable.max(Traversable.scala:108)
	at com.nvidia.spark.rapids.tool.profiling.Analysis.$anonfun$getMaxTaskInputSizeBytes$1(Analysis.scala:335)
	at scala.collection.immutable.List.map(List.scala:293)
	at com.nvidia.spark.rapids.tool.profiling.Analysis.getMaxTaskInputSizeBytes(Analysis.scala:316)
	at com.nvidia.spark.rapids.tool.profiling.Profiler.com$nvidia$spark$rapids$tool$profiling$Profiler$$processApps(Profiler.scala:315)
	at com.nvidia.spark.rapids.tool.profiling.Profiler$ProfileProcessThread$1.run(Profiler.scala:230)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
```

**Expected behavior**
Check that the sequence is nonEmpty before getting max.

### PR DESCRIPTION
**Pull request by amahussein**
Signed-off-by: Ahmed Hussein (amahussein) <a@ahussein.me>

fixes #6707, fixes #6895

**Changed made for 6707:**

The objective is to do smallest change that allows testing autotuner without loading entire eventlog.  

- Added new classes in ApplicationSummaryInfo.scala: `AppSummaryInfoBaseProvider`, and `SingleAppSummaryInfoProvider`
- Change the constructor of AutoTuner, and make necessary changes in the implementation to avoid direct access of `ApplicationSummaryInfo`.
- Added a couple of new unit tests. We should be able to build on top of those unit tests when we start evaluating the heuristics.

**Changed made for 6895:**

- check that the sequence is non empty before calling max in Analysis.scala
- if it is empty, return `getMaxTaskInputSizeBytes` returns 0.

### CODE REVIEW

**Comment by amahussein** [comment]
build

**Comment by amahussein** [comment]
build

**Comment by amahussein** [comment]
build

**Comment by amahussein** [comment]
build

File: tools/src/test/scala/com/nvidia/spark/rapids/tool/profiling/AutoTunerSuite.scala [hunk 1]
```diff
@@ -99,7 +125,8 @@ class AutoTunerSuite extends FunSuite with BeforeAndAfterEach with Logging {
 
   test("Load cluster properties with CPU cores 0") {
     val dataprocWorkerInfo = buildWorkerInfoAsString(None, Some(0))
-    val autoTuner: AutoTuner = AutoTuner.buildAutoTunerFromProps(dataprocWorkerInfo, None)
+    val autoTuner: AutoTuner = AutoTuner.buildAutoTunerFromProps(dataprocWorkerInfo,
+      getMockInfoProvider)
```
Comment (tgravescs) [inline tools/src/test/scala/com/nvidia/spark/rapids/tool/profiling/AutoTunerSuite.scala:129]: only concern here is that the getMockInfoProvider is not used and its not completely obvious here at first look.  Can we perhaps create one that is UnusedProvider

File: tools/src/test/scala/com/nvidia/spark/rapids/tool/profiling/AutoTunerSuite.scala [hunk 2]
```diff
@@ -423,9 +460,159 @@ class AutoTunerSuite extends FunSuite with BeforeAndAfterEach with Logging {
...
+        "spark.executor.instances" -> "1",
+        "spark.sql.shuffle.partitions" -> "200",
+        "spark.sql.files.maxPartitionBytes" -> "1g",
+        "spark.task.resource.gpu.amount" -> "0.0625",
+        "spark.rapids.memory.pinnedPool.size" -> "5g",
+        "spark.rapids.sql.enabled" -> "true",
+        "spark.rapids.sql.concurrentGpuTasks" -> "4")
+
+    val autoTuner: AutoTuner = AutoTuner.buildAutoTunerFromProps(dataprocWorkerInfo,
+      getMockInfoProvider(8126464.0, Seq(0), Seq(0.004), logEventsProps,
```
Comment (tgravescs) [inline tools/src/test/scala/com/nvidia/spark/rapids/tool/profiling/AutoTunerSuite.scala:494]: this is a bit hard to read can you put this on separate line:

mockProvider = getMock.... and pass mockProvider into buildAutoTuner

File: tools/src/test/scala/com/nvidia/spark/rapids/tool/profiling/AutoTunerSuite.scala [hunk 3]
```diff
@@ -423,9 +460,159 @@ class AutoTunerSuite extends FunSuite with BeforeAndAfterEach with Logging {
...
           |""".stripMargin
-    val autoTuner: AutoTuner = AutoTuner.buildAutoTunerFromProps(dataprocWorkerInfo, None)
+    val autoTuner: AutoTuner = AutoTuner.buildAutoTunerFromProps(dataprocWorkerInfo,
+      getMockInfoProvider)
     val (properties, comments) = autoTuner.getRecommendedProperties()
     val autoTunerOutput = Profiler.getAutoTunerResultsAsString(properties, comments)
     assert(expectedResults == autoTunerOutput)
   }
+
+  test("test with app mock 001") {
```
Comment (tgravescs) [inline tools/src/test/scala/com/nvidia/spark/rapids/tool/profiling/AutoTunerSuite.scala:470]: what is this testing?

File: tools/src/test/scala/com/nvidia/spark/rapids/tool/profiling/AutoTunerSuite.scala [hunk 4]
```diff
@@ -423,9 +460,159 @@ class AutoTunerSuite extends FunSuite with BeforeAndAfterEach with Logging {
...
+          |Comments:
+          |- 'spark.executor.memoryOverhead' was not set.
+          |- 'spark.executor.memoryOverhead' must be set if using 'spark.rapids.memory.pinnedPool.size
+          |- 'spark.sql.adaptive.enabled' should be enabled for better performance.
+          |""".stripMargin
+    // scalastyle:on line.size.limit
+    assert(expectedResults == autoTunerOutput)
+  }
+
+  test("test with app mock 002") {
```
Comment (tgravescs) [inline tools/src/test/scala/com/nvidia/spark/rapids/tool/profiling/AutoTunerSuite.scala:520]: same here would like to know what htis is trying to test

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/ApplicationSummaryInfo.scala [hunk 5]
```diff
@@ -39,3 +39,90 @@ case class ApplicationSummaryInfo(
     val wholeStage: Seq[WholeStageCodeGenResults],
     val maxTaskInputBytesRead: Seq[SQLMaxTaskInputSizes],
     val appLogPath: Seq[AppLogPathProfileResults])
+
+trait AppInfoPropertyGetter {
+  def isAppDefined: Boolean = true
```
[the remark this answers is not in the document]
Comment (amahussein) [reply tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/ApplicationSummaryInfo.scala]: You are right. I forgot to remove it before pushing.

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/ApplicationSummaryInfo.scala [hunk 6]
```diff
@@ -39,3 +39,90 @@ case class ApplicationSummaryInfo(
...
+  def getMaxInput: Double
+}
+
+abstract class AppSummaryInfoBaseProvider extends AppInfoPropertyGetter
+  with AppInfoSqlTaskAggMetricsVisitor
+  with AppInfoSQLMaxTaskInputSizes {
+  override def getSparkProperty(propKey: String): Option[String] = None
+  override def getRapidsProperty(propKey: String): Option[String] = None
+  override def getProperty(propKey: String): Option[String]
+  override def getSparkVersion: Option[String]
```
[the remark this answers is not in the document]
Comment (amahussein) [reply tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/ApplicationSummaryInfo.scala]: I filled the empty definitions, and removed the abstract keyword.

File: tools/src/test/scala/com/nvidia/spark/rapids/tool/profiling/AutoTunerSuite.scala (hunk 3 above)
Comment (amahussein) [reply tools/src/test/scala/com/nvidia/spark/rapids/tool/profiling/AutoTunerSuite.scala]: Changed the description and added a comment above the method.

File: tools/src/test/scala/com/nvidia/spark/rapids/tool/profiling/AutoTunerSuite.scala (hunk 4 above)
Comment (amahussein) [reply tools/src/test/scala/com/nvidia/spark/rapids/tool/profiling/AutoTunerSuite.scala]: changed that too.

File: tools/src/test/scala/com/nvidia/spark/rapids/tool/profiling/AutoTunerSuite.scala (hunk 1 above)
Comment (amahussein) [reply tools/src/test/scala/com/nvidia/spark/rapids/tool/profiling/AutoTunerSuite.scala]: I changed the name of the method to be "getUnusedProvider" which constructs an object of `AppSummaryInfoBaseProvider`

**Comment by amahussein** [comment]
build

**Review by tgravescs — APPROVED** [review]

**Pull request merged**